File: server/sonar-server/src/main/java/org/sonar/server/webhook/WebHooks.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.server.webhook;

import java.util.Objects;
import java.util.function.Supplier;
import javax.annotation.CheckForNull;
import javax.annotation.Nullable;
import org.sonar.api.config.Configuration;

import static java.util.Objects.requireNonNull;

public interface WebHooks {
  /**
   * Tells whether any webHook is configured at all for the specified {@link Configuration}.
   *
   * <p>
   * This can be used to not do consuming operations before calling
   * {@link #sendProjectAnalysisUpdate(Configuration, Analysis, Supplier)}
   */
  boolean isEnabled(Configuration configuration);

  /**
   * Calls all WebHooks configured in the specified {@link Configuration} for the specified analysis with the
   * {@link WebhookPayload} provided by the specified Supplier.
   */
  void sendProjectAnalysisUpdate(Configuration configuration, Analysis analysis, Supplier<WebhookPayload> payloadSupplier);

  final class Analysis {
    private final String projectUuid;
    private final String ceTaskUuid;
    private final String analysisUuid;

    public Analysis(String projectUuid, @Nullable String analysisUuid, @Nullable  String ceTaskUuid) {
      this.projectUuid = requireNonNull(projectUuid, "projectUuid can't be null");
      this.analysisUuid = analysisUuid;
      this.ceTaskUuid = ceTaskUuid;
    }

    public String getProjectUuid() {
      return projectUuid;
    }

    @CheckForNull
    public String getCeTaskUuid() {
      return ceTaskUuid;
    }

    @CheckForNull
    public String getAnalysisUuid() {
      return analysisUuid;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      Analysis analysis = (Analysis) o;
      return Objects.equals(projectUuid, analysis.projectUuid) &&
        Objects.equals(ceTaskUuid, analysis.ceTaskUuid) &&
        Objects.equals(analysisUuid, analysis.analysisUuid);
    }

    @Override
    public int hashCode() {
      return Objects.hash(projectUuid, ceTaskUuid, analysisUuid);
    }

    @Override
    public String toString() {
      return "Analysis{" +
        "projectUuid='" + projectUuid + '\'' +
        ", ceTaskUuid='" + ceTaskUuid + '\'' +
        '}';
    }
  }
}

File: server/sonar-web/src/main/js/api/mocks/data/webhooks.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { ComponentQualifier } from '~sonar-aligned/types/component';
import { mockComponent } from '../../../helpers/mocks/component';
import { mockWebhook, mockWebhookDelivery } from '../../../helpers/mocks/webhook';
import { WebhookDelivery, WebhookResponse } from '../../../types/webhook';
import {
  WEBHOOK_GLOBAL_1,
  WEBHOOK_GLOBAL_1_LATEST_DELIVERY_ID,
  WEBHOOK_GLOBAL_2,
  WEBHOOK_PROJECT_1,
  WEBHOOK_PROJECT_1_1,
  WEBHOOK_PROJECT_1_2,
} from './ids';

export const webhookProject = mockComponent({
  key: WEBHOOK_PROJECT_1,
  qualifier: ComponentQualifier.Project,
});

const globalWebhook1Deliveries = [
  mockWebhookDelivery({
    at: '2019-06-24T09:45:52+0200',
    id: WEBHOOK_GLOBAL_1_LATEST_DELIVERY_ID,
  }),
  ...Array.from({ length: 15 }).map((_, i) =>
    mockWebhookDelivery({
      id: `global-webhook-1-delivery-${i}`,
      at: `2019-06-${(23 - i).toString().padStart(2, '0')}T09:45:52+0200`,
      httpStatus: i % 2 === 0 ? 200 : undefined,
      success: i % 2 === 0,
      durationMs: 1000 + i * 100,
    }),
  ),
];

const project1Webhook1Deliveries = [
  mockWebhookDelivery({
    id: 'project-1-delivery-1',
    at: '2019-06-24T09:45:52+0200',
  }),
];

export const deliveries = [...globalWebhook1Deliveries, ...project1Webhook1Deliveries];

export const webhooks: Record<string, Array<WebhookResponse>> = {
  global: [
    mockWebhook({
      key: WEBHOOK_GLOBAL_1,
      name: 'Global webhook 1',
      url: 'https://example.com/1',
      latestDelivery: globalWebhook1Deliveries[0],
    }),
    mockWebhook({
      key: WEBHOOK_GLOBAL_2,
      name: 'Global webhook 2',
      hasSecret: true,
      url: 'https://example.com/2',
    }),
  ],
  [webhookProject.key]: [
    mockWebhook({
      key: WEBHOOK_PROJECT_1_1,
      name: 'Project 1 webhook 1',
      url: 'https://example.com/1',
      latestDelivery: project1Webhook1Deliveries[0],
    }),
    mockWebhook({
      key: WEBHOOK_PROJECT_1_2,
      name: 'Project 1 webhook 2',
      url: 'https://example.com/2',
    }),
  ],
};

export function mockFullWebhookList(project?: string): Array<WebhookResponse> {
  return cloneDeep(webhooks[project ?? 'global'] ?? []);
}

export function mockFullWebhookDeliveries(webhook?: string): Array<WebhookDelivery> {
  return cloneDeep(webhook === WEBHOOK_GLOBAL_1 ? globalWebhook1Deliveries : []);
}
